Source/dafny.msbuild/DafnyVersionCheckTask.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Task = Microsoft.Build.Utilities.Task;
+
+namespace DafnyMSBuild
+{
+    /**
+     * A custom Task for checking that the provided Dafny CLI is of a compatible version.
+     */
+    public class DafnyVersionCheckTask : Task
+    {
+        public string DafnyBinariesDir { get; set; }
+
+        [Required]
+        public string DafnyExecutablePath { get; set; }
+
+        // Either a Dafny version number, like "2.3.0.10506", or the empty string.
+        public string DafnyVersion { get; set; }
+
+        // Either a Git commit hash, or the empty string.
+        public string DafnyVersionCommit { get; set; }
+
+        // Either a Git tag name, or the empty string.
+        public string DafnyVersionTag { get; set; }
+
+        public override bool Execute()
+        {
+            var foundDafnyVersion = GetDafnyExecutableVersion(DafnyExecutablePath);
+
+            var enforceVersion = !string.IsNullOrEmpty(DafnyVersion);
+            var enforceCommit = !string.IsNullOrEmpty(DafnyVersionCommit);
+            var enforceTag = !string.IsNullOrEmpty(DafnyVersionTag);
+            if (!(enforceVersion || enforceCommit || enforceTag)) return Success(foundDafnyVersion);
+            if (new[] {enforceVersion, enforceCommit, enforceTag}.Count(b => b) > 1) {
+                return Fail("The DafnyVersion, DafnyVersionCommit, and DafnyVersionTag properties are mutually exclusive");
+            }
+
+            if (enforceVersion) return DafnyVersion == foundDafnyVersion
+                ? Success(foundDafnyVersion)
+                : Fail($"Expected Dafny version \"{DafnyVersion}\", found \"{foundDafnyVersion}\"");
```

Comment:
nit: This is a bit over my threshold for single-line cleverness. :) I would put the if body in braces.